// arrow-admin-ui/config.js
module.exports = {
	// default values for the API and UI prefixes.
	// Used if in modes like build/test where API Builder isn't available to get the values.
	prefix: '/console',
	apiprefix: '/adminapi/v1',
	apiDocPrefix: '/apidoc',
	endpointsEnabled: true,
	docSummaryUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/index.html',
	docManageSpecs: 'https://docs.axway.com/bundle/api-builder/page/docs/guide_openapi/managing_apidocs/index.html',
	docUpgradeEndpoints: 'https://docs.axway.com/bundle/api-builder/page/docs/guide_openapi/upgrading/index.html',
	docConfUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/developer_guide/project/configuration/project_configuration/index.html',
	docCredentialUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/developer_guide/credentials/index.html',
	docFlowUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/developer_guide/flows/index.html',
	docTriggerUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/developer_guide/flows/flow_triggers/index.html',
	docModelUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/developer_guide/models/index.html',
	docUIUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/index.html',
	docComponentsUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/developer_guide/plugins/index.html',
	docPluginsSDKUrl: 'https://docs.axway.com/bundle/api-builder/page/docs/developer_guide/sdk/index.html',
	feedbackUrl: 'https://amplifyideas.axway.com/?project=APIB'
};
