// @axway/api-builder-runtime/deprecations.js
module.exports = {
	url: 'https://docs.axway.com/bundle/api-builder/page/docs/deprecations/index.html#',
	features: {
		apiDocPrefix: {
			id: 'D002',
			message: 'admin.apiDocPrefix has been deprecated, use apidoc.prefix instead'
		},
		disableAPIDoc: {
			id: 'D003',
			message: 'admin.disableAPIDoc has been deprecated'
		},
		apidocDisabled: {
			id: 'D054',
			message: 'apidoc.disabled has been deprecated'
		},
		perURLAuthentication: {
			id: 'D010',
			message: 'Config settings related to authentication have been deprecated (APIKeyAuthType, and APIKeyAuthPlugin), use accessControl for API security instead'
		},
		matchURL: {
			id: 'D010',
			message: 'The configured authentication plugin is using using a deprecated method \'matchURL\' for applying security to paths and is potentially insecure'
		},
		apiBuilderWeb: {
			id: 'D011',
			message: 'API Builder Web feature (routes in web/routes folder) has been deprecated and rendering engines are not registered by default. Please use another custom approach for your web application needs'
		},
		codeblock: {
			id: 'D028',
			message: 'Codeblocks have been deprecated, use @axway/api-builder-plugin-fn-javascript flow-node or use the axway-flow-sdk to build a custom flow-node'
		},
		port: {
			id: 'D038',
			message: '\'port\' configuration setting has been deprecated, use \'http.port\' instead'
		},
		nodeSupport: {
			id: 'D039',
			message: `Using Node.js ${process.version} with API Builder has been deprecated since it is no longer supported or maintained. Use a more recent Node.js version instead`
		},
		nodeSupport14: {
			id: 'D055',
			message: 'Using Node.js v14 with API Builder has been deprecated and support will be dropped in API Builder v5'
		},
		bodyParser: {
			id: 'D046',
			message: 'Using config.bodyParser has been deprecated'
		},
		busboy: {
			id: 'D047',
			message: 'Using config.busboy has been deprecated, use config.limits to limit multipart/form-data part size instead'
		},
		undefinedMultipartPartSize: {
			id: 'D048',
			message: 'Not setting config.limits.multipartPartSize, or using undefined as the value will result in inconsistent behavior when making multipart/form-data requests. This behavior has been deprecated. We recommend configuring a reasonable limit'
		},
		disableServerBanner: {
			id: 'D051',
			message: 'Using config.disableServerBanner has been deprecated. Instead, set config.http.headers.server to false to disable the header'
		},
		runtimePeerDependency: {
			id: 'D052',
			message: 'Using dependencies with a peerDependency on @axway/api-builder-runtime has been deprecated and may cause issues when using npm 7 or higher',
			usageMessage: 'The following dependencies should be updated'
		}
	},
	flags: {
		enableModelsWithNoPrimaryKey: {
			id: 'D004',
			message: 'Deprecation: Models that do not have a primary will behave differently in a future release. In the future release, for models that do not have a primary key, the delete, findAndModify, findByID, upsert, and update APIs and endpoints will not be generated, the model flow-node will not have delete, findAndModify, findByID, upsert, or update methods, and the Create API will no longer return a location header. For information on how be prepared for the change and start using the new behavior now, refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/models_no_primary_key/index.html'
		},
		usePrimaryKeyType: {
			id: 'D005',
			message: 'Deprecation: Models with primary keys that are not strings will behave differently in a future release. In the future release, Model IDs will be typed based on the database primary key instead of being hard-coded as a string. For information on how be prepared for the change and start using the new behavior now, refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/use_primary_key_type/index.html'
		},
		exitOnPluginFailure: {
			id: 'D006',
			message: 'Deprecation: The startup of an API Builder server with plugins that fail to load correctly will behave differently in a future release. In the future release, the server will fail to start when a plugin fails to load correctly. For information on how be prepared for the change and start using the new behavior now, refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/exit_on_plugin_failure/index.html'
		},
		enableAliasesInCompositeOperators: {
			id: 'D007',
			message: 'Deprecation: Composite model queries will behave differently in a future release. In the future release, composite model query will support aliased fields for all comparison operators - $lt, $gt, $lte, $gte, $in, $nin, $ne and $eq. For information on how be prepared for the change and start using the new behavior now, refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/aliases_in_comparason_operators/index.html'
		},
		enableMemoryConnectorLike: {
			id: 'D008',
			message: 'Deprecation: The Memory connector will behave differently in a future release. In the future release, the memory connector will support the $like comparison operator. For information on how be prepared for the change and start using the new behavior now, refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/memory_connector_like/index.html'
		},
		enableScopedConfig: {
			id: 'D009',
			message: 'Deprecation: API Builder may pass the entire config to a plugin when it is being loaded. In the future release, a plugin will only receive the config relevant to that plugin. For information on how be prepared for the change and start using the new behavior now, refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/scoped_config/index.html'
		},
		enableModelNameEncoding: {
			id: 'D012',
			message: 'Deprecation: Model names having characters not in `A-Z a-z 0-9 - _ . ~` will behave differently in a future release.	In the future release, model names will be percent encoded in their URI as per RFC-3986. For information on how be prepared for the change and start using the new behavior now, refer to https://bundle/api-builder/page/docs/deprecations/model_name_encoding/index.html'
		},
		enableNullModelFields: {
			id: 'D013',
			message: 'Deprecation: Responses from Models with null fields be different in a future release.	In the future release, Models can return null fields in the response instead of hiding the field. For information on how be prepared for the change and start using the new behavior now, refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/null_model_fields/index.html'
		},
		enableModelNameEncodingInSwagger: {
			id: 'D014',
			message: 'Deprecation: Model names having characters not in `A-Z a-z 0-9 - _ . ~` will behave differently in a future release.	In the future release, model names used in Swagger will be percent encoded in their URI as per RFC-3986, refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/model_name_encoding_swagger/index.html'
		},
		enableModelNameEncodingWithConnectorSlash: {
			id: 'D015',
			message: 'Deprecation: Auto-generated APIs created from Models which are prefixed with their connector name (i.e. oracle/user) will behave differently in a future release. In the future release, these APIs will have a literal slash instead of \'%2F\', refer to https://docs.axway.com/bundle/api-builder/page/docs/deprecations/model_name_encoding_connector_slash/index.html'
		},
		enableOverrideEndpointContentType: {
			id: 'D042',
			message: 'Deprecation: Endpoints always respond with Content-Type \'application/json\' and the body is always JSON encoded, and will behave differently in a future release. In the future release, endpoints will use Content-Type from the flow (if one was set) and will not always JSON encode the body, instead, it will send the body appropriate to the body type: https://docs.axway.com/bundle/api-builder/page/docs/deprecations/endpoint_content_type/index.html'
		},
		enableModelErrorOutputs: {
			id: 'D043',
			message: 'Deprecation: Model flow-nodes that encounter an unexpected error will no longer abort, and will behave differently in a future release.	In the future release, model flow-nodes will have an Error output and database errors will resolve that output instead, refer to: https://docs.axway.com/bundle/api-builder/page/docs/deprecations/model_error_outputs/index.html'
		},
		exitOnSwaggerSchemaValidationError: {
			id: 'D045',
			message: 'Deprecation: Validation errors when loading JSON schemas will cause the service to terminate. Additionally, The service\'s Swagger API docs are now validated and will cause the service to terminate if found to be invalid. For information on how be prepared for the change and start using the new behavior now, refer to: https://docs.axway.com/bundle/api-builder/page/docs/deprecations/exit_on_swagger_schema_validation_error/index.html'
		},
		enableLoggingOfLevel: {
			id: 'D049',
			message: 'Deprecation: Logs will include the log level in each output in a future release. For information on how be prepared for the change and start using the new behavior now, refer to: https://docs.axway.com/bundle/api-builder/page/docs/deprecations/logging_of_level/index.html'
		},
		enableStrictBodyPayloads: {
			id: 'D050',
			message: 'Deprecation: HTTP methods GET and HEAD will ignore payload body. For information on how be prepared for the change and start using the new behavior now, refer to: https://docs.axway.com/bundle/api-builder/page/docs/deprecations/strict_body_payloads/index.html'
		},
		disableEndpoints: {
			id: 'D053',
			message: 'Deprecation: Endpoints have been deprecated. For information on how to upgrade to the OpenAPI flow-trigger plugin, refer to: https://docs.axway.com/bundle/api-builder/page/docs/guide_openapi/upgrading/index.html'
		}
	}
};
