// arrow-admin-api/src/endpoints/modelendpoints/createAPI.js
const { isAutoGenerated } = require('../util');

module.exports = ({ operationId, schemaRef, model, encodedModelName }) => {
	return {
		swagger: '2.0',
		info: {
			title: `${model.name}`,
			version: '1.0.0'
		},
		paths: {
			[`/${encodedModelName}`]: {
				post: {
					operationId,
					tags: [
						`${model.name}`
					],
					summary: `Create a ${model.name}`,
					consumes: [
						'application/json'
					],
					produces: [
						'application/json'
					],
					parameters: [
						{
							name: 'data',
							in: 'body',
							description: 'Create body',
							schema: {
								$ref: (!model.hasPrimaryKey() || isAutoGenerated(model))
									? schemaRef.model : schemaRef.full
							}
						}
					],
					responses: {
						201: {
							description: 'The create succeeded.',
							...(model.hasPrimaryKey() ? {
								headers: {
									Location: {
										description: 'The URL to the newly created instance.',
										type: 'string'
									}
								}
							} : {})
						},
						400: {
							description: 'Bad request.'
						},
						404: {
							description: 'No results were found.'
						},
						500: {
							description: 'Something went wrong during the request; check out the logs on your server.'
						}
					}
				}
			}
		}
	};
};
