// arrow-admin-api/src/endpoints/modelendpoints/distinctAPI.js
const { arrowQueryOptions } = require('../../util');

module.exports = ({ operationId, model, encodedModelName, schemaRef }) => {
	const additionalParameters = arrowQueryOptions.filter(a => a.name !== 'sel' && a.name !== 'unsel');
	return {
		swagger: '2.0',
		info: {
			title: `${model.name}`,
			version: '1.0.0'
		},
		paths: {
			[`/${encodedModelName}/distinct/{field}`]: {
				get: {
					operationId,
					tags: [
						`${model.name}`
					],
					summary: `Gets a distinct set of \`field\` value(s) from the ${model.name} collection`,
					consumes: [],
					produces: [
						'application/json'
					],
					parameters: [
						{
							in: 'path',
							name: 'field',
							description: 'The field name that must be distinct.',
							required: true,
							type: 'string'
						},
						...additionalParameters
					],
					responses: {
						200: {
							description: 'Distinct fields response',
							schema: {
								type: 'array',
								items: {
									$ref: schemaRef.modelEx
								},
								uniqueItems: true
							}
						},
						400: {
							description: 'Bad request.'
						},
						404: {
							description: 'No results were found.'
						},
						500: {
							description: 'Something went wrong during the request; check out the logs on your server.'
						}
					}
				}
			}
		}
	};
};
