// arrow-admin-api/src/endpoints/modelendpoints/findByIDAPI.js

module.exports = ({ operationId, schemaRef, model, encodedModelName }) => {
	return {
		swagger: '2.0',
		info: {
			title: `${model.name}`,
			version: '1.0.0'
		},
		paths: {
			[`/${encodedModelName}/{${model.getPrimaryKeyName()}}`]: {
				get: {
					operationId,
					tags: [
						`${model.name}`
					],
					summary: `Find one ${model.name}`,
					consumes: [],
					produces: [
						'application/json'
					],
					parameters: [
						{
							name: model.getPrimaryKeyName(),
							in: 'path',
							description: `The ${model.name} ID`,
							required: true,
							type: schemaRef.pkType
						}
					],
					responses: {
						200: {
							description: 'The find succeeded, and the results are available.',
							schema: {
								$ref: schemaRef.full
							}
						},
						400: {
							description: 'Bad request.'
						},
						404: {
							description: 'No results were found.'
						},
						500: {
							description: 'Something went wrong during the request; check out the logs on your server.'
						}
					}
				}
			}
		}
	};
};
