// arrow-admin-api/src/endpoints/modelendpoints/updateAPI.js

module.exports = ({ operationId, schemaRef, model, encodedModelName }) => {
	return {
		swagger: '2.0',
		info: {
			title: `${model.name}`,
			version: '1.0.0'
		},
		paths: {
			[`/${encodedModelName}/{${model.getPrimaryKeyName()}}`]: {
				put: {
					operationId,
					tags: [
						`${model.name}`
					],
					summary: `Update a ${model.name}`,
					consumes: [
						'application/json'
					],
					produces: [
						'application/json'
					],
					parameters: [
						{
							name: model.getPrimaryKeyName(),
							in: 'path',
							description: `The ${model.name} ID`,
							required: true,
							type: schemaRef.pkType
						},
						{
							name: 'data',
							in: 'body',
							description: 'Update body',
							schema: {
								$ref: schemaRef.model
							}
						}
					],
					responses: {
						204: {
							description: 'The update succeeded.'
						},
						400: {
							description: 'Bad request.'
						},
						404: {
							description: 'No results were found.'
						},
						500: {
							description: 'Something went wrong during the request; check out the logs on your server.'
						}
					}
				}
			}
		}
	};
};
