// arrow-admin-api/src/endpoints/modelendpoints/upsertAPI.js
module.exports = ({ operationId, schemaRef, model, encodedModelName }) => {
	return {
		swagger: '2.0',
		info: {
			title: `${model.name}`,
			version: '1.0.0'
		},
		paths: {
			[`/${encodedModelName}/upsert`]: {
				post: {
					operationId,
					tags: [
						`${model.name}`
					],
					summary: `Create or update a ${model.name}`,
					consumes: [
						'application/json'
					],
					produces: [
						'application/json'
					],
					parameters: [
						{
							name: 'data',
							in: 'body',
							description: 'Upsert body',
							schema: {
								$ref: schemaRef.full
							}
						}
					],
					responses: {
						201: {
							description: 'The upsert succeeded, and resulted in an insert.',
							headers: {
								Location: {
									description: 'The URL to the newly created instance.',
									type: 'string'
								}
							}
						},
						204: {
							description: 'The upsert succeeded, and resulted in an update.'
						},
						400: {
							description: 'Bad request.'
						},
						404: {
							description: 'No results were found.'
						},
						500: {
							description: 'Something went wrong during the request; check out the logs on your server.'
						}
					}
				}
			}
		}
	};
};
