// arrow-admin-api/src/endpoints/modelflows/countAPIV2.js
const { getShortName } = require('../util');
const { arrowQueryParameter } = require('../../util');

const countAPI = (name, schemaRef, model, options) => ({
	schemaVersion: '4',
	info: {
		name,
		description: `Count ${model.name}`
	},
	parameter: {
		properties: {
			...(options.withParams && {
				params: {
					type: 'object',
					description: 'The parameters from the endpoint.',
					properties: {
						where: arrowQueryParameter.where
					}
				}
			}),
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			...(options.withParams ? [ 'params' ] : []),
			'request',
			'config',
			'env'
		]
	},
	start: 'model.count',
	nodes: {
		'model.count': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'count',
			name: `Count ${getShortName(model, true)}`,
			parameters: [
				{
					name: 'where',
					value: `$.${options.withParams ? 'params' : 'request.query'}.where`
				}
			],
			outputs: {
				next: {
					context: '$.count',
					routes: [ 'response.success' ]
				},
				error: {
					context: '$.error',
					routes: [ 'response.error' ]
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Success (HTTP 200)',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '200',
					type: 'number'
				},
				{
					name: 'body',
					value: '$.count'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		},
		'response.error': {
			type: 'nodehandler://axway-flow/http',
			name: 'Bad Request (HTTP 400)',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '400',
					type: 'number'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

module.exports = countAPI;
