// arrow-admin-api/src/endpoints/modelflows/createAPIV2.js
const { getShortName, isAutoGenerated } = require('../util');

const createNoPk = (name, schemaRef, model, options) => ({
	schemaVersion: '4',
	info: {
		name,
		description: `Create ${model.name}`
	},
	parameter: {
		properties: {
			...(options.withParams && {
				params: {
					type: 'object',
					description: 'The parameters from the endpoint.',
					properties: {
						data: {
							$ref: schemaRef.model,
							description: `The ${model.name} to create`
						}
					},
					required: [
						'data'
					]
				}
			}),
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			...(options.withParams ? [ 'params' ] : []),
			'request',
			'config',
			'env'
		]
	},
	start: 'model.create',
	nodes: {
		'model.create': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'create',
			name: `Create ${getShortName(model)}`,
			parameters: [
				{
					name: 'data',
					value: `$.${options.withParams ? 'params.data' : 'request.body'}`
				}
			],
			outputs: {
				next: {
					context: '$.model',
					routes: [ 'response.success' ]
				},
				error: {
					context: '$.error',
					routes: [ 'response.error' ]
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set success response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '201',
					type: 'number'
				},
				{
					name: 'headers',
					value: '$.headers'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		},
		'response.error': {
			type: 'nodehandler://axway-flow/http',
			name: 'Bad Request (HTTP 400)',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '400',
					type: 'number'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

const createWithPk = (name, schemaRef, model, options) => ({
	schemaVersion: '4',
	info: {
		name,
		description: `Create ${model.name}`
	},
	parameter: {
		properties: {
			...(options.withParams && {
				params: {
					type: 'object',
					description: 'The parameters from the endpoint.',
					properties: {
						data: {
							$ref: isAutoGenerated(model) ? schemaRef.model : schemaRef.full,
							description: `The ${model.name} to create`
						}
					},
					required: [
						'data'
					]
				}
			}),
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			...(options.withParams ? [ 'params' ] : []),
			'request',
			'config',
			'env'
		]
	},
	start: 'model.create',
	nodes: {
		'model.create': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'create',
			name: `Create ${getShortName(model)}`,
			parameters: [
				{
					name: 'data',
					value: `$.${options.withParams ? 'params.data' : 'request.body'}`
				}
			],
			outputs: {
				next: {
					context: '$.model',
					routes: [ 'set.headers.location' ]
				},
				error: {
					context: '$.error',
					routes: [ 'response.error' ]
				}
			}
		},
		'set.headers.location': {
			type: 'nodehandler://axway-flow/setcontext',
			name: 'Set Headers',
			method: 'setcontext',
			parameters: [
				{
					name: 'value',
					type: 'jsonpath',
					value: `$.model['${model.getPrimaryKeyName()}']`
				}
			],
			outputs: {
				next: {
					routes: [
						'response.success'
					],
					context: '$.headers.location'
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Success (HTTP 200)',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '201',
					type: 'number'
				},
				{
					name: 'headers',
					value: '$.headers'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		},
		'response.error': {
			type: 'nodehandler://axway-flow/http',
			name: 'Bad Request (HTTP 400)',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '400',
					type: 'number'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

const createAPI = (name, schemaRef, model, options) => {
	return model.hasPrimaryKey()
		? createWithPk(name, schemaRef, model, options)
		: createNoPk(name, schemaRef, model, options);
};

module.exports = createAPI;
