// arrow-admin-api/src/endpoints/modelflows/queryAPIV1.js
const { getShortName } = require('../util');
const { arrowQueryParameter } = require('../../util');

const queryAPI = (name, schemaRef, model, options) => ({
	schemaVersion: '4',
	info: {
		name,
		description: `Query ${model.name}`
	},
	parameter: {
		properties: {
			...(options.withParams && {
				params: {
					type: 'object',
					description: 'The parameters from the endpoint.',
					properties: {
						...arrowQueryParameter
					}
				}
			}),
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			...(options.withParams ? [ 'params' ] : []),
			'request',
			'config',
			'env'
		]
	},
	start: 'model.query',
	nodes: {
		'model.query': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'query',
			name: `Query ${getShortName(model, true)}`,
			parameters: [
				{
					name: 'limit',
					value: `$.${options.withParams ? 'params' : 'request.query'}.limit`
				},
				{
					name: 'skip',
					value: `$.${options.withParams ? 'params' : 'request.query'}.skip`
				},
				{
					name: 'where',
					value: `$.${options.withParams ? 'params' : 'request.query'}.where`
				},
				{
					name: 'order',
					value: `$.${options.withParams ? 'params' : 'request.query'}.order`
				},
				{
					name: 'sel',
					value: `$.${options.withParams ? 'params' : 'request.query'}.sel`
				},
				{
					name: 'unsel',
					value: `$.${options.withParams ? 'params' : 'request.query'}.unsel`
				}
			],
			outputs: {
				next: {
					context: '$.models',
					routes: [ 'response.success' ]
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '200',
					type: 'number'
				},
				{
					name: 'body',
					value: '$.models'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

module.exports = queryAPI;
