// arrow-admin-api/src/endpoints/modelflows/upsertAPIV2.js
const { getShortName } = require('../util');

const upsertAPI = (name, schemaRef, model, options) => ({
	schemaVersion: '4',
	info: {
		name,
		description: `Upsert ${model.name}`
	},
	parameter: {
		properties: {
			...(options.withParams && {
				params: {
					type: 'object',
					description: 'The parameters from the endpoint.',
					properties: {
						data: {
							$ref: schemaRef.full,
							description: `The ${model.name} to upsert`
						}
					},
					required: [
						'data'
					]
				}
			}),
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			...(options.withParams ? [ 'params' ] : []),
			'request',
			'config',
			'env'
		]
	},
	start: 'model.upsert',
	nodes: {
		'model.upsert': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'upsert',
			name: `Upsert ${getShortName(model)}`,
			parameters: [
				{
					name: 'data',
					value: `$.${options.withParams ? 'params.data' : 'request.body'}`
				}
			],
			outputs: {
				insert: {
					context: '$.model',
					routes: [ 'set.headers.location' ]
				},
				update: {
					context: '$.model',
					routes: [ 'update.success' ]
				},
				error: {
					context: '$.error',
					routes: [ 'response.error' ]
				}
			}
		},
		'set.headers.location': {
			type: 'nodehandler://axway-flow/setcontext',
			method: 'setcontext',
			name: 'Set Location',
			parameters: [
				{
					name: 'value',
					value: '$.model.id'
				}
			],
			outputs: {
				next: {
					routes: [ 'insert.success' ],
					context: '$.headers.location'
				}
			}
		},
		'insert.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Created (HTTP 201)',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '201',
					type: 'number'
				},
				{
					name: 'headers',
					value: '$.headers'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		},
		'update.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Updated (HTTP 204)',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '204',
					type: 'number'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		},
		'response.error': {
			type: 'nodehandler://axway-flow/http',
			name: 'Bad Request (HTTP 400)',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '400',
					type: 'number'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

module.exports = upsertAPI;
