// arrow-admin-api/src/project/ProjectController.js
const path = require('path');

/**
 * The controller implementing the actions for interacting with the service.
 * @public
 */
class ProjectController {
	/**
     * Create a ProjectController.
     * @param {object} ctx - The context for this controller.
     * @param {express} ctx.app - The express application.
     * @param {string} ctx.prefix - The prefix bound to.
     * @param {APIBuilder} ctx.apibuilder - The API Builder server.
     * @param {Swagger} ctx.swagger - openapi-doc instance.
     */
	constructor(ctx) {
		this.ctx = ctx;
	}

	/**
     * Get the service summary.
	 *
     * @return {object} summary - The service summary.
     * @return {string} summary.name - The service name.
     * @return {string} summary.version - The service version.
     * @return {string} summary.author - The service author.
     * @return {string} summary.description - The service description.
     * @return {string} summary.license - The service license.
     * @return {string} summary.apikey - The service apikey.
     * @return {string} summary.apiPrefixSecurity - The security applied to the apiPrefix.
     */
	getSummary() {
		return new Promise((resolve) => {
			const config = this.ctx.apibuilder.config;

			let summary = {};

			const packagePath = path.join(config.dir, 'package');

			try {
				// eslint-disable-next-line security/detect-non-literal-require
				const pkg = require(packagePath);
				let author;
				/**
				 * Parses the value of author if it's an object and turns it
				 * into a string.
				 *
				 * Example:
				 * "author": {
				 *		"name": "Banana",
				 *		"email": "banana@axway.com",
				 *		"url": "https://axway.com"
				 * }
				 *
				 * Will be transformed to:
				 *  Banana <banana@axway.com> (https://axway.com)
				 */
				if (typeof pkg.author === 'object') {
					const { name, email, url } = pkg.author;
					const parts = [ name ];
					if (email) {
						parts.push(`<${email}>`);
					}
					if (url) {
						parts.push(`(${url})`);
					}
					author = parts.join(' ');
				} else {
					author = pkg.author;
				}
				summary = {
					name: pkg.name,
					version: pkg.version,
					author,
					description: pkg.description,
					license: pkg.license,
					apikey: config.apikey,
					apiPrefixSecurity: config.accessControl.apiPrefixSecurity
				};
			} catch (importError) {
				// Project didn't have a package.json at the given location, resolve with an empty
				// object
			}

			resolve(summary);
		});
	}
}

module.exports = ProjectController;
